File: db/npc_trade.go
// Package db NPC 行腳／兜售：與討論 002「不必 assignment 亦可賣貨換鎂」對齊。
package db

import (
	"encoding/json"
	"fmt"
	"math/rand"
)

// NpcStreetSellOne 在街頭賣出背包中第一筆數量≥1 的物品 1 件，換取少量鎂；供 IntentTrade 到達效果。
func NpcStreetSellOne(entityID, roomID string) bool {
	c, err := GetEntity(entityID)
	if err != nil || c == nil {
		return false
	}
	if c.Inventory == "" || c.Inventory == "[]" {
		return false
	}
	var entries []InventoryEntry
	if err := json.Unmarshal([]byte(c.Inventory), &entries); err != nil {
		return false
	}
	for _, e := range entries {
		if e.Qty <= 0 || e.ItemID == "" {
			continue
		}
		if err := RemoveFromInventory(entityID, e.ItemID, 1); err != nil {
			continue
		}
		mag := 4 + rand.Intn(10)
		_ = AddMagnesium(entityID, mag)
		LogNPCEvent(entityID, EvtTrade, fmt.Sprintf("在%s兜售賣出%s，得%d鎂", roomID, e.ItemID, mag))
		AdjustDisposition(entityID, DispTrade)
		return true
	}
	return false
}

File: src/db/npc_trade.rs
//! NPC 街頭兜售（對齊 Go `db/npc_trade.go`）。

use rand::Rng;

use super::{
    add_magnesium, adjust_disposition, get_entity, log_npc_event, remove_from_inventory, DISP_TRADE, EVT_TRADE,
};

/// 賣出背包中第一筆 `qty≥1` 的物品 1 件換鎂；成功回 `true`（對齊 Go `NpcStreetSellOne`）。
pub fn npc_street_sell_one(entity_id: &str, room_id: &str, at: i64) -> bool {
    let Ok(Some(ch)) = get_entity(entity_id) else {
        return false;
    };
    if ch.inventory.is_empty() || ch.inventory == "[]" {
        return false;
    }
    #[derive(serde::Deserialize)]
    struct InvRow {
        #[serde(rename = "item_id")]
        item_id: String,
        qty: i32,
    }
    let entries: Vec<InvRow> = serde_json::from_str(&ch.inventory).unwrap_or_default();
    for e in &entries {
        if e.qty <= 0 || e.item_id.is_empty() {
            continue;
        }
        if remove_from_inventory(entity_id, &e.item_id, 1).is_err() {
            continue;
        }
        let mut rng = rand::rng();
        let mag = 4 + rng.random_range(0..10);
        let _ = add_magnesium(entity_id, mag);
        let payload = format!("在{room_id}兜售賣出{}，得{mag}鎂", e.item_id);
        log_npc_event(at, entity_id, EVT_TRADE, &payload);
        let _ = adjust_disposition(entity_id, DISP_TRADE);
        return true;
    }
    false
}
